src/main/java/graphql/schema/idl/TypeDefinitionRegistry.java

```diff
@@ -0,0 +1,135 @@
+
+import graphql.GraphQLError;
+import graphql.language.Definition;
+import graphql.language.ScalarTypeDefinition;
+import graphql.language.SchemaDefinition;
+import graphql.language.Type;
+import graphql.language.TypeDefinition;
+import graphql.language.TypeExtensionDefinition;
+import graphql.language.TypeName;
+import graphql.schema.idl.errors.SchemaRedefinitionError;
+import graphql.schema.idl.errors.TypeRedefinitionError;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Optional.ofNullable;
+
+/**
+ * A {@link TypeDefinitionRegistry} contains the set of type definitions that come from compiling
+ * a graphql schema definition file via {@link SchemaCompiler#compile(String)}
+ */
+public class TypeDefinitionRegistry {
+
+    private final Map<String, ScalarTypeDefinition> scalarTypes = new LinkedHashMap<>();
+    private final Map<String, TypeExtensionDefinition> typeExtensions = new LinkedHashMap<>();
+    private final Map<String, TypeDefinition> types = new LinkedHashMap<>();
+    private SchemaDefinition schema;
+
+    public Optional<GraphQLError> add(Definition definition) {
+        if (definition instanceof TypeExtensionDefinition) {
+            TypeExtensionDefinition newEntry = (TypeExtensionDefinition) definition;
+            return define(typeExtensions, typeExtensions, newEntry);
+        } else if (definition instanceof ScalarTypeDefinition) {
+            ScalarTypeDefinition newEntry = (ScalarTypeDefinition) definition;
+            return define(scalarTypes, scalarTypes, newEntry);
+        } else if (definition instanceof TypeDefinition) {
+            TypeDefinition newEntry = (TypeDefinition) definition;
+            return define(types, types, newEntry);
+        } else if (definition instanceof SchemaDefinition) {
+            SchemaDefinition newSchema = (SchemaDefinition) definition;
+            if (schema != null) {
+                return Optional.of(new SchemaRedefinitionError(this.schema, newSchema));
+            } else {
+                schema = newSchema;
+            }
+        }
+        return Optional.empty();
+    }
+
+    private <T extends TypeDefinition> Optional<GraphQLError> define(Map<String, T> source, Map<String, T> target, T newEntry) {
+        String name = newEntry.getName();
+
+        T olderEntry = source.get(name);
+        if (olderEntry != null) {
+            return Optional.of(handleReDefinition(olderEntry, newEntry));
+        } else {
+            target.put(name, newEntry);
+        }
+        return Optional.empty();
+    }
+
+    public Optional<List<GraphQLError>> merge(TypeDefinitionRegistry typeRegistry) {
+        List<GraphQLError> errors = new ArrayList<>();
+
+        Map<String, TypeDefinition> tempTypes = new LinkedHashMap<>();
+        typeRegistry.types.values().forEach(newEntry -> define(this.types, tempTypes, newEntry).ifPresent(errors::add));
```

Comment:
OK added and extra test to confirm.  Circular dependencies are ok and handled ok.

eg  

    type Author {  
        post : [Post]
    }

and in another file

     type Post {
       author : Author;
    }

The code only cares at this stage for top level types

References to types are resolved later in the SchemaGenerator